reco_utils/evaluation/spark_evaluation.py

```diff
@@ -470,3 +479,457 @@ def _get_relevant_items_by_timestamp(
+            Ziegler, McNee, Konstan and Lausen, 2005.
+        """
+        if self.df_intralist_similarity is None:
+            pairs = self._get_pairwise_items(df=df)
+            similarity_df = self._get_cosine_similarity()
+            # Fillna(0) is needed in the cases where similarity_df does not have an entry for a pair of items.
+            # e.g. i1 and i2 have never occurred together.
+            self.df_intralist_similarity = (
+                pairs.join(similarity_df, on=["i1", "i2"], how="left")
+                .fillna(0)
+                .filter(F.col("i1") != F.col("i2"))
+                .groupBy(self.col_user)
+                .agg(F.mean(self.sim_col).alias("avg_il_sim"))
+                .select(self.col_user, "avg_il_sim")
+            )
+        return self.df_intralist_similarity
+
+    def user_diversity(self):
+        """Calculate average diversity of recommendations for each user.
+        The metric definition is based on formula (3) in the following reference:
+
+        :Citation:
+
+            Y.C. Zhang, D.Ó. Séaghdha, D. Quercia and T. Jambor, Auralist:
+            introducing serendipity into music recommendation, WSDM 2012
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: A dataframe with the following columns: col_user, user_diversity.
+        """
+        if self.df_user_diversity is None:
+            self.df_intralist_similarity = self._get_intralist_similarity(self.reco_df)
+            self.df_user_diversity = (
+                self.df_intralist_similarity.withColumn(
+                    "user_diversity", 1 - F.col("avg_il_sim")
+                )
+                .select(self.col_user, "user_diversity")
+                .orderBy(self.col_user)
+            )
+        return self.df_user_diversity
+
+    def diversity(self):
+        """Calculate average diversity of recommendations across all users.
+
+        Returns:
+            float: diversity.
+        """
+        if self.df_diversity is None:
+            self.df_user_diversity = self.user_diversity()
+            self.df_diversity = self.df_user_diversity.agg(
+                {"user_diversity": "mean"}
+            ).first()[0]
+        return self.df_diversity
+
+    # Novelty metrics
+    def item_novelty(self):
+        """Calculate novelty for each item. Novelty is computed as the minus logarithm of
+        (number of interactions with item / total number of interactions). The definition of the metric
+        is based on the following reference using the choice model (eqs. 1 and 6):
+
+        :Citation:
+
+            P. Castells, S. Vargas, and J. Wang, Novelty and diversity metrics for recommender systems:
+            choice, discovery and relevance, ECIR 2011
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: A dataframe with the following columns: col_item, item_novelty.
+        """
+        if self.df_item_novelty is None:
+            n_records = self.train_df.count()
```

Comment:
p(i) as defined here is (# times item i is recommended across all users) / (# all interactions in training data)
i would expect p(i) to be more like (# times item i is recommended over all users) / (# all recommendations)

Reply:
No, actually novelty of each item (`item_novelty`) is calculated on the historical data (`train_df`). Novelty of the recommendations (`novelty`) is restricted to those items in `reco_df`. 
I follow Castells et al. Item novelty is based on p(i) from eq. (1). As they say,p(i) represents the probability that ݅i is observed. It is basically the popularity of the item, so it has to come from data about user choice, not from what you recommend to them (after all you may recommend items that are not popular to them).

Reply:
do we have separate metrics for novelty assessed on historical data versus recommendations from a model? We should make this distinction clear to users, because I would expect evaluation metrics to apply to the model not the training data, just like the rating and ranking metrics.

Reply:
So, the novelty metric depends on the recommendations and hence the model. See https://github.com/microsoft/recommenders/blob/a66d452531e503cef9a5d495cb941058b5654ead/reco_utils/evaluation/spark_evaluation.py#L789
At the notebook, the novelties of ALS and random are not the same.

Reply:
ok, i get it now. thanks